add ROCm 7

phueper opened the pull request:
update to ROCm 7.1

i am very unsure about the change in CMakePresets, just basically copied the ROCm 6 presets and renamed it for consistency. Only tested on my gfx1151 based AMD RYZEN AI MAX+ 395 w/ Radeon 8060S

Commit:
move missing dependency resolution

- insteead of adding in the final image, add it as part of the ollama build

Commit:
add ROCm 7 presets to CMakePresets.json

very unsure about this, i just copied ROCm 6 for now, works for my gfx1151 card for now

Commit:
update Docker build to ROCm 7.1

Commit:
add rocroller dependency for ROCm 7

phueper commented:
Note: this is based on and includes the changes of #26, so probably #26 should be handled first (it just moves the missing dependency from an ubuntu dependency to ollama cmake output)

rjmalagon commented:
I will branch this for a rocm 7.1 image container image, parallel to the "optm " releases, because 7.1 only supports very recent APUs.

rjmalagon commented:
I will add this on https://github.com/rjmalagon/ollama-linux-amd-apu/tree/apu-optimizer-rocm7

rjmalagon commented:
I don't have the required hardware to test it. If you update your patches, I will accept your pull request on that branch.

rjmalagon commented:
You may test it ghcr.io/rjmalagon/ollama-linux-amd-apu:optm-rocm7-latest
Unfortunately, fails on GFX1030-level hardware.

phueper commented:
interesting ... according to https://rocm.docs.amd.com/projects/install-on-linux/en/docs-7.1.0/reference/system-requirements.html#supported-gpus the gfx1030 should be supported  :/

phueper commented:
ghcr.io/rjmalagon/ollama-linux-amd-apu:optm-rocm7-latest

works for me :+1:

Commit:
remove unnecessary find line

Commit:
build and include both rocm_v6 and rocm_v7 in rocm FLAVOR build

so we can use rocm_v7 for Hardware that is supported and rocm_v6 for Hardware that isnt

phueper commented:
@rjmalagon  i updated my changes... they are now building both ROCm 6.4.4 and ROCm 7.1.0 into the `rocm` FLAVOR image ... for me this allows ollama to autoselect rocm_v7 for my hardware, if i set OLLAMA_LLM_LIBRARY to "rocm_v6" it the runs using rocm_v6 ... in theory this could work for your setup when it detects that gfx1030 doesnt work with rocm_v7 that it falls back to rocm_v6  ... could you try this please? :)

jammsen commented:
Hey @phueper i finally got it running and spend the last 2 days when i wasnt ill to full-automate it.

My setup does ROCm 7.1.0 on Proxmox -> LXC -> Docker -> ollama:rocm, works great but i think that rocm is 6.3.3 - https://github.com/ollama/ollama/blob/main/Dockerfile#L6

Source see here:
https://github.com/jammsen/proxmox-setup-scripts

If you want i can test yours if you give me the details and your image.
Everything else, PVE-Host + LXC-Guest runs ROCm 7.1.0

I also dont have a 395+ i have only a mere HX PRO 370 😄 but that way we can test 2 devices.
I also can use HSA_OVERRIDE_GFX_VERSION=11.5.1 or 11.5.0 which results in gfx1151 (yours) or gfx1150(mine)

phueper commented:
> If you want i can test yours if you give me the details and your image.

Hi, i pushed an image based on the current state of this PR ghcr.io/phueper/ollama-linux-amd-apu:optm-rocm_6_7

feel free to test it, it includes rocm_v6 (6.4.4), rocm_v7 (7.1.0) and vulkan libraries, you can either have it autodetect or manually select a version using

OLLAMA_LLM_LIBRARY="rocm_v6|rocm_v7|vulkan" 

environment settings. i am not using other environment switches for my setup, besides OLLAMA_FLASH_ATTENTION=true (and OLLAMA_DEBUG=1 to see which libraries are loaded and which ones are chosen :) )

phueper commented:
@rjmalagon  thx for merging, i'm curious though ... does it mean you tested with your gfx1030 and the auto selection worked ? so the fallback to rocm_v6 was used ?

rjmalagon commented:
No, I tested a ROCm 7 only build.
I will test this multi ROCm build later , and I will try to build images around it, further testing if stable enough to distribute.

jammsen commented:
> > If you want i can test yours if you give me the details and your image.
> 
> Hi, i pushed an image based on the current state of this PR ghcr.io/phueper/ollama-linux-amd-apu:optm-rocm_6_7
> 
> feel free to test it, it includes rocm_v6 (6.4.4), rocm_v7 (7.1.0) and vulkan libraries, you can either have it autodetect or manually select a version using
> 
> OLLAMA_LLM_LIBRARY="rocm_v6|rocm_v7|vulkan"
> 
> environment settings. i am not using other environment switches for my setup, besides OLLAMA_FLASH_ATTENTION=true (and OLLAMA_DEBUG=1 to see which libraries are loaded and which ones are chosen :) )

Testing it right now with my setup:

```bash
docker run -d \
  --name ollama-rocm7-gpu \
  -p 11434:11434 \
  --device /dev/kfd \
  --device /dev/dri \
  -e OLLAMA_LLM_LIBRARY=rocm_v7 \
  -e OLLAMA_FLASH_ATTENTION=true \
  -e OLLAMA_DEBUG=1 \
  -e HSA_OVERRIDE_GFX_VERSION=11.5.1  \
  -e HSA_ENABLE_SDMA=0 \
  --group-add video \
  --cap-add=SYS_PTRACE \
  --security-opt seccomp=unconfined \
  --ipc=host \
  -v ollama-models:/root/.ollama \
  ghcr.io/phueper/ollama-linux-amd-apu:optm-rocm_6_7 && docker logs -f ollama-rocm7-gpu
```

Sadly the GH image download is a 300KB/s this will take ages -.-

jammsen commented:
```
Unable to find image 'ghcr.io/phueper/ollama-linux-amd-apu:optm-rocm_6_7' locally
optm-rocm_6_7: Pulling from phueper/ollama-linux-amd-apu
485e128c8cc7: Downloading [=========>                                         ]   7.16MB/36.77MB
674d19983d33: Downloading [===>                                               ]  6.487MB/86.02MB
2beaa600699c: Downloading [=============================>                     ]  7.193MB/12.15MB
157e9d4c8def: Waiting
```

jammsen commented:
```
time=2025-11-04T22:35:00.482Z level=INFO source=routes.go:1524 msg="server config" env="map[CUDA_VISIBLE_DEVICES: GGML_VK_VISIBLE_DEVICES: GPU_DEVICE_ORDINAL: HIP_VISIBLE_DEVICES: HSA_OVERRIDE_GFX_VERSION:11.5.1 HTTPS_PROXY: HTTP_PROXY: NO_PROXY: OLLAMA_CONTEXT_LENGTH:4096 OLLAMA_DEBUG:DEBUG OLLAMA_FLASH_ATTENTION:true OLLAMA_GPU_OVERHEAD:0 OLLAMA_HOST:http://0.0.0.0:11434 OLLAMA_INTEL_GPU:false OLLAMA_KEEP_ALIVE:5m0s OLLAMA_KV_CACHE_TYPE: OLLAMA_LLM_LIBRARY:rocm_v7 OLLAMA_LOAD_TIMEOUT:5m0s OLLAMA_MAX_LOADED_MODELS:0 OLLAMA_MAX_QUEUE:512 OLLAMA_MODELS:/root/.ollama/models OLLAMA_MULTIUSER_CACHE:false OLLAMA_NEW_ENGINE:false OLLAMA_NOHISTORY:false OLLAMA_NOPRUNE:false OLLAMA_NUM_PARALLEL:1 OLLAMA_ORIGINS:[http://localhost https://localhost http://localhost:* https://localhost:* http://127.0.0.1 https://127.0.0.1 http://127.0.0.1:* https://127.0.0.1:* http://0.0.0.0 https://0.0.0.0 http://0.0.0.0:* https://0.0.0.0:* app://* file://* tauri://* vscode-webview://* vscode-file://*] OLLAMA_REMOTES:[ollama.com] OLLAMA_SCHED_SPREAD:false ROCR_VISIBLE_DEVICES: http_proxy: https_proxy: no_proxy:]"
time=2025-11-04T22:35:00.482Z level=INFO source=images.go:522 msg="total blobs: 5"
time=2025-11-04T22:35:00.483Z level=INFO source=images.go:529 msg="total unused blobs removed: 0"
[GIN-debug] [WARNING] Creating an Engine instance with the Logger and Recovery middleware already attached.

[GIN-debug] [WARNING] Running in "debug" mode. Switch to "release" mode in production.
 - using env:   export GIN_MODE=release
 - using code:  gin.SetMode(gin.ReleaseMode)

[GIN-debug] HEAD   /                         --> github.com/ollama/ollama/server.(*Server).GenerateRoutes.func1 (5 handlers)
[GIN-debug] GET    /                         --> github.com/ollama/ollama/server.(*Server).GenerateRoutes.func2 (5 handlers)
[GIN-debug] HEAD   /api/version              --> github.com/ollama/ollama/server.(*Server).GenerateRoutes.func3 (5 handlers)
[GIN-debug] GET    /api/version              --> github.com/ollama/ollama/server.(*Server).GenerateRoutes.func4 (5 handlers)
[GIN-debug] POST   /api/pull                 --> github.com/ollama/ollama/server.(*Server).PullHandler-fm (5 handlers)
[GIN-debug] POST   /api/push                 --> github.com/ollama/ollama/server.(*Server).PushHandler-fm (5 handlers)
[GIN-debug] HEAD   /api/tags                 --> github.com/ollama/ollama/server.(*Server).ListHandler-fm (5 handlers)
[GIN-debug] GET    /api/tags                 --> github.com/ollama/ollama/server.(*Server).ListHandler-fm (5 handlers)
[GIN-debug] POST   /api/show                 --> github.com/ollama/ollama/server.(*Server).ShowHandler-fm (5 handlers)
[GIN-debug] DELETE /api/delete               --> github.com/ollama/ollama/server.(*Server).DeleteHandler-fm (5 handlers)
[GIN-debug] POST   /api/me                   --> github.com/ollama/ollama/server.(*Server).WhoamiHandler-fm (5 handlers)
[GIN-debug] POST   /api/signout              --> github.com/ollama/ollama/server.(*Server).SignoutHandler-fm (5 handlers)
[GIN-debug] DELETE /api/user/keys/:encodedKey --> github.com/ollama/ollama/server.(*Server).SignoutHandler-fm (5 handlers)
[GIN-debug] POST   /api/create               --> github.com/ollama/ollama/server.(*Server).CreateHandler-fm (5 handlers)
[GIN-debug] POST   /api/blobs/:digest        --> github.com/ollama/ollama/server.(*Server).CreateBlobHandler-fm (5 handlers)
[GIN-debug] HEAD   /api/blobs/:digest        --> github.com/ollama/ollama/server.(*Server).HeadBlobHandler-fm (5 handlers)
[GIN-debug] POST   /api/copy                 --> github.com/ollama/ollama/server.(*Server).CopyHandler-fm (5 handlers)
[GIN-debug] GET    /api/ps                   --> github.com/ollama/ollama/server.(*Server).PsHandler-fm (5 handlers)
[GIN-debug] POST   /api/generate             --> github.com/ollama/ollama/server.(*Server).GenerateHandler-fm (5 handlers)
[GIN-debug] POST   /api/chat                 --> github.com/ollama/ollama/server.(*Server).ChatHandler-fm (5 handlers)
[GIN-debug] POST   /api/embed                --> github.com/ollama/ollama/server.(*Server).EmbedHandler-fm (5 handlers)
[GIN-debug] POST   /api/embeddings           --> github.com/ollama/ollama/server.(*Server).EmbeddingsHandler-fm (5 handlers)
[GIN-debug] POST   /v1/chat/completions      --> github.com/ollama/ollama/server.(*Server).ChatHandler-fm (6 handlers)
[GIN-debug] POST   /v1/completions           --> github.com/ollama/ollama/server.(*Server).GenerateHandler-fm (6 handlers)
[GIN-debug] POST   /v1/embeddings            --> github.com/ollama/ollama/server.(*Server).EmbedHandler-fm (6 handlers)
[GIN-debug] GET    /v1/models                --> github.com/ollama/ollama/server.(*Server).ListHandler-fm (6 handlers)
[GIN-debug] GET    /v1/models/:model         --> github.com/ollama/ollama/server.(*Server).ShowHandler-fm (6 handlers)
time=2025-11-04T22:35:00.483Z level=INFO source=routes.go:1577 msg="Listening on [::]:11434 (version 0.0.0)"
time=2025-11-04T22:35:00.483Z level=DEBUG source=sched.go:120 msg="starting llm scheduler"
time=2025-11-04T22:35:00.483Z level=INFO source=runner.go:76 msg="discovering available GPUs..."
time=2025-11-04T22:35:00.484Z level=INFO source=server.go:400 msg="starting runner" cmd="/usr/bin/ollama runner --ollama-engine --port 37257"
time=2025-11-04T22:35:00.484Z level=DEBUG source=server.go:401 msg=subprocess PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin HSA_OVERRIDE_GFX_VERSION=11.5.1 HSA_ENABLE_SDMA=0 OLLAMA_LLM_LIBRARY=rocm_v7 OLLAMA_FLASH_ATTENTION=true OLLAMA_DEBUG=1 LD_LIBRARY_PATH=/usr/lib/ollama:/usr/lib/ollama/rocm_v7:/usr/local/nvidia/lib:/usr/local/nvidia/lib64:/usr/lib/ollama:/usr/lib/ollama/rocm OLLAMA_HOST=0.0.0.0:11434 OLLAMA_LIBRARY_PATH=/usr/lib/ollama:/usr/lib/ollama/rocm_v7
time=2025-11-04T22:35:00.582Z level=DEBUG source=runner.go:471 msg="bootstrap discovery took" duration=98.855851ms OLLAMA_LIBRARY_PATH="[/usr/lib/ollama /usr/lib/ollama/rocm_v7]" extra_envs=map[]
time=2025-11-04T22:35:00.582Z level=DEBUG source=runner.go:91 msg="skipping available library at users request" requested=rocm_v7 libDir=/usr/lib/ollama/vulkan
time=2025-11-04T22:35:00.582Z level=DEBUG source=runner.go:91 msg="skipping available library at users request" requested=rocm_v7 libDir=/usr/lib/ollama/rocm_v6
time=2025-11-04T22:35:00.582Z level=DEBUG source=runner.go:120 msg="evluating which if any devices to filter out" initial_count=1
time=2025-11-04T22:35:00.582Z level=DEBUG source=runner.go:132 msg="verifying GPU is supported" library=/usr/lib/ollama/rocm_v7 description="AMD Radeon 890M Graphics" compute=gfx1151 id=0 pci_id=0000:c8:00.0
time=2025-11-04T22:35:00.583Z level=INFO source=server.go:400 msg="starting runner" cmd="/usr/bin/ollama runner --ollama-engine --port 33165"
time=2025-11-04T22:35:00.583Z level=DEBUG source=server.go:401 msg=subprocess PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin HSA_OVERRIDE_GFX_VERSION=11.5.1 HSA_ENABLE_SDMA=0 OLLAMA_LLM_LIBRARY=rocm_v7 OLLAMA_FLASH_ATTENTION=true OLLAMA_DEBUG=1 LD_LIBRARY_PATH=/usr/lib/ollama:/usr/lib/ollama/rocm_v7:/usr/local/nvidia/lib:/usr/local/nvidia/lib64:/usr/lib/ollama:/usr/lib/ollama/rocm OLLAMA_HOST=0.0.0.0:11434 OLLAMA_LIBRARY_PATH=/usr/lib/ollama:/usr/lib/ollama/rocm_v7 GGML_CUDA_INIT=1 ROCR_VISIBLE_DEVICES=0
time=2025-11-04T22:35:01.176Z level=DEBUG source=runner.go:471 msg="bootstrap discovery took" duration=594.001827ms OLLAMA_LIBRARY_PATH="[/usr/lib/ollama /usr/lib/ollama/rocm_v7]" extra_envs="map[GGML_CUDA_INIT:1 ROCR_VISIBLE_DEVICES:0]"
time=2025-11-04T22:35:01.177Z level=DEBUG source=runner.go:41 msg="GPU bootstrap discovery took" duration=693.592951ms
time=2025-11-04T22:35:01.177Z level=INFO source=types.go:42 msg="inference compute" id=0 filtered_id=0 library=ROCm compute=gfx1151 name=ROCm0 description="AMD Radeon 890M Graphics" libdirs=ollama,rocm_v7 driver=70125.42 pci_id=0000:c8:00.0 type=iGPU total="96.0 GiB" available="95.8 GiB"
```

```
total duration:       4m18.764490753s
load duration:        5.560176382s
prompt eval count:    170 token(s)
prompt eval duration: 597.874172ms
prompt eval rate:     284.34 tokens/s
eval count:           4724 token(s)
eval duration:        4m11.517853153s
eval rate:            18.78 tokens/s
root@ollama-docker-amd-230:~# docker exec ollama-rocm7-gpu ollama run --verbose gpt-oss:20b "Design a scalable web application architecture for an e-commerce platform that needs to handle 10,000 concurrent users during peak shopping periods. Your architecture should include a detailed database design with proper normalization, effective caching strategies using Redis or Memcached, load balancers, CDN integration, payment processing systems, inventory management, user authentication and authorization, session management, and monitoring tools. Explain the key components, their interactions, data flow between services, and how the system maintains performance and reliability under high traffic loads while ensuring secure transactions."
```

6.3.3 had 18.63 TPS, so its withing margin of error / run to run variances.

How can i find out which ROCm version running in this image?
apt list --installed or ls -al /opt/ or something similar dont return anything.

jammsen commented:
Is ROCm 7 -> 7.0.1 or 7.0.2 or 7.10?

phueper commented:
> How can i find out which ROCm version running in this image?

https://github.com/rjmalagon/ollama-linux-amd-apu/pull/27/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R6

```
ARG ROCM6VERSION=6.4.4
ARG ROCM7VERSION=7.1
```

since you are forcing it to rocm_v7  it is using ROCm 7.1.0

phueper commented:
btw ... testing the image on my  ` AMD Ryzen AI 5 340 w/ Radeon 840M`  Laptop i get **slightly** higher performance (as in ~14 T/s vs ~11 T/s) by using the `vulkan` driver instead of `rocm_v7` or `rocm_v6`

jammsen commented:
I can later test vulkan, need to do work first :)

rjmalagon commented:
Mmm, @phueper  the dual ROCM builds ok, but runtime selection is not smart enough to pick ROCM 6 when ROCM 7 fails "after detection." On GFX1030 level hardware, runtime detection works on ROCM v7, dies miserably at model load.

Runtime selection via environment variable works, but you can force only one at a time. 

I will try tinkering a bit before releasing the container images; maybe I will limit ROCM 7 targets to GFX1100 level onwards.

jammsen commented:
I added now this for more testing:
https://github.com/jammsen/proxmox-setup-scripts/blob/main/docker-compose-testing-examples/docker-compose.yml

Im starting to test Vulkan next to watching series, will post an image when i have a few models together 😎 
Here is a sneak peak from an old version of the docker run setup.

Vulkan-Testing:

```
root@ollama-docker-amd-230:~# docker run -d \
  --name ollama-vulkan-gpu \
  -p 11434:11434 \
  --device /dev/kfd \
  --device /dev/dri \
  -e OLLAMA_LLM_LIBRARY=vulkan \
  -e OLLAMA_FLASH_ATTENTION=true \
  -e OLLAMA_DEBUG=1 \
  -e HSA_OVERRIDE_GFX_VERSION=11.5.1  \
  -e HSA_ENABLE_SDMA=0 \
  --group-add video \
  --cap-add=SYS_PTRACE \
  --security-opt seccomp=unconfined \
  --ipc=host \
  -v ollama-models:/root/.ollama \
  ghcr.io/phueper/ollama-linux-amd-apu:optm-rocm_6_7 && docker logs -f ollama-vulkan-gpu
1c24a8fbc491f1bb0f75129a6a94b0fd324c65850b836ea7dc10f40a5952dd35
time=2025-11-05T18:51:54.611Z level=INFO source=routes.go:1524 msg="server config" env="map[CUDA_VISIBLE_DEVICES: GGML_VK_VISIBLE_DEVICES: GPU_DEVICE_ORDINAL: HIP_VISIBLE_DEVICES: HSA_OVERRIDE_GFX_VERSION:11.5.1 HTTPS_PROXY: HTTP_PROXY: NO_PROXY: OLLAMA_CONTEXT_LENGTH:4096 OLLAMA_DEBUG:DEBUG OLLAMA_FLASH_ATTENTION:true OLLAMA_GPU_OVERHEAD:0 OLLAMA_HOST:http://0.0.0.0:11434 OLLAMA_INTEL_GPU:false OLLAMA_KEEP_ALIVE:5m0s OLLAMA_KV_CACHE_TYPE: OLLAMA_LLM_LIBRARY:vulkan OLLAMA_LOAD_TIMEOUT:5m0s OLLAMA_MAX_LOADED_MODELS:0 OLLAMA_MAX_QUEUE:512 OLLAMA_MODELS:/root/.ollama/models OLLAMA_MULTIUSER_CACHE:false OLLAMA_NEW_ENGINE:false OLLAMA_NOHISTORY:false OLLAMA_NOPRUNE:false OLLAMA_NUM_PARALLEL:1 OLLAMA_ORIGINS:[http://localhost https://localhost http://localhost:* https://localhost:* http://127.0.0.1 https://127.0.0.1 http://127.0.0.1:* https://127.0.0.1:* http://0.0.0.0 https://0.0.0.0 http://0.0.0.0:* https://0.0.0.0:* app://* file://* tauri://* vscode-webview://* vscode-file://*] OLLAMA_REMOTES:[ollama.com] OLLAMA_SCHED_SPREAD:false ROCR_VISIBLE_DEVICES: http_proxy: https_proxy: no_proxy:]"
time=2025-11-05T18:51:54.614Z level=INFO source=images.go:522 msg="total blobs: 10"
time=2025-11-05T18:51:54.614Z level=INFO source=images.go:529 msg="total unused blobs removed: 0"
[GIN-debug] [WARNING] Creating an Engine instance with the Logger and Recovery middleware already attached.

[GIN-debug] [WARNING] Running in "debug" mode. Switch to "release" mode in production.
 - using env:   export GIN_MODE=release
 - using code:  gin.SetMode(gin.ReleaseMode)

[GIN-debug] HEAD   /                         --> github.com/ollama/ollama/server.(*Server).GenerateRoutes.func1 (5 handlers)
[GIN-debug] GET    /                         --> github.com/ollama/ollama/server.(*Server).GenerateRoutes.func2 (5 handlers)
[GIN-debug] HEAD   /api/version              --> github.com/ollama/ollama/server.(*Server).GenerateRoutes.func3 (5 handlers)
[GIN-debug] GET    /api/version              --> github.com/ollama/ollama/server.(*Server).GenerateRoutes.func4 (5 handlers)
[GIN-debug] POST   /api/pull                 --> github.com/ollama/ollama/server.(*Server).PullHandler-fm (5 handlers)
[GIN-debug] POST   /api/push                 --> github.com/ollama/ollama/server.(*Server).PushHandler-fm (5 handlers)
[GIN-debug] HEAD   /api/tags                 --> github.com/ollama/ollama/server.(*Server).ListHandler-fm (5 handlers)
[GIN-debug] GET    /api/tags                 --> github.com/ollama/ollama/server.(*Server).ListHandler-fm (5 handlers)
[GIN-debug] POST   /api/show                 --> github.com/ollama/ollama/server.(*Server).ShowHandler-fm (5 handlers)
[GIN-debug] DELETE /api/delete               --> github.com/ollama/ollama/server.(*Server).DeleteHandler-fm (5 handlers)
[GIN-debug] POST   /api/me                   --> github.com/ollama/ollama/server.(*Server).WhoamiHandler-fm (5 handlers)
[GIN-debug] POST   /api/signout              --> github.com/ollama/ollama/server.(*Server).SignoutHandler-fm (5 handlers)
[GIN-debug] DELETE /api/user/keys/:encodedKey --> github.com/ollama/ollama/server.(*Server).SignoutHandler-fm (5 handlers)
[GIN-debug] POST   /api/create               --> github.com/ollama/ollama/server.(*Server).CreateHandler-fm (5 handlers)
[GIN-debug] POST   /api/blobs/:digest        --> github.com/ollama/ollama/server.(*Server).CreateBlobHandler-fm (5 handlers)
[GIN-debug] HEAD   /api/blobs/:digest        --> github.com/ollama/ollama/server.(*Server).HeadBlobHandler-fm (5 handlers)
[GIN-debug] POST   /api/copy                 --> github.com/ollama/ollama/server.(*Server).CopyHandler-fm (5 handlers)
[GIN-debug] GET    /api/ps                   --> github.com/ollama/ollama/server.(*Server).PsHandler-fm (5 handlers)
[GIN-debug] POST   /api/generate             --> github.com/ollama/ollama/server.(*Server).GenerateHandler-fm (5 handlers)
[GIN-debug] POST   /api/chat                 --> github.com/ollama/ollama/server.(*Server).ChatHandler-fm (5 handlers)
[GIN-debug] POST   /api/embed                --> github.com/ollama/ollama/server.(*Server).EmbedHandler-fm (5 handlers)
[GIN-debug] POST   /api/embeddings           --> github.com/ollama/ollama/server.(*Server).EmbeddingsHandler-fm (5 handlers)
[GIN-debug] POST   /v1/chat/completions      --> github.com/ollama/ollama/server.(*Server).ChatHandler-fm (6 handlers)
[GIN-debug] POST   /v1/completions           --> github.com/ollama/ollama/server.(*Server).GenerateHandler-fm (6 handlers)
[GIN-debug] POST   /v1/embeddings            --> github.com/ollama/ollama/server.(*Server).EmbedHandler-fm (6 handlers)
[GIN-debug] GET    /v1/models                --> github.com/ollama/ollama/server.(*Server).ListHandler-fm (6 handlers)
[GIN-debug] GET    /v1/models/:model         --> github.com/ollama/ollama/server.(*Server).ShowHandler-fm (6 handlers)
time=2025-11-05T18:51:54.615Z level=INFO source=routes.go:1577 msg="Listening on [::]:11434 (version 0.0.0)"
time=2025-11-05T18:51:54.615Z level=DEBUG source=sched.go:120 msg="starting llm scheduler"
time=2025-11-05T18:51:54.617Z level=INFO source=runner.go:76 msg="discovering available GPUs..."
time=2025-11-05T18:51:54.617Z level=DEBUG source=runner.go:91 msg="skipping available library at users request" requested=vulkan libDir=/usr/lib/ollama/rocm_v6
time=2025-11-05T18:51:54.617Z level=DEBUG source=runner.go:91 msg="skipping available library at users request" requested=vulkan libDir=/usr/lib/ollama/rocm_v7
time=2025-11-05T18:51:54.618Z level=INFO source=server.go:400 msg="starting runner" cmd="/usr/bin/ollama runner --ollama-engine --port 45625"
time=2025-11-05T18:51:54.618Z level=DEBUG source=server.go:401 msg=subprocess PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin OLLAMA_FLASH_ATTENTION=true OLLAMA_DEBUG=1 HSA_OVERRIDE_GFX_VERSION=11.5.1 HSA_ENABLE_SDMA=0 OLLAMA_LLM_LIBRARY=vulkan LD_LIBRARY_PATH=/usr/lib/ollama:/usr/lib/ollama/vulkan:/usr/local/nvidia/lib:/usr/local/nvidia/lib64:/usr/lib/ollama:/usr/lib/ollama/rocm OLLAMA_HOST=0.0.0.0:11434 OLLAMA_LIBRARY_PATH=/usr/lib/ollama:/usr/lib/ollama/vulkan
time=2025-11-05T18:51:54.786Z level=DEBUG source=runner.go:471 msg="bootstrap discovery took" duration=169.042537ms OLLAMA_LIBRARY_PATH="[/usr/lib/ollama /usr/lib/ollama/vulkan]" extra_envs=map[]
time=2025-11-05T18:51:54.786Z level=DEBUG source=runner.go:120 msg="evluating which if any devices to filter out" initial_count=1
time=2025-11-05T18:51:54.786Z level=DEBUG source=runner.go:132 msg="verifying GPU is supported" library=/usr/lib/ollama/vulkan description="AMD Radeon 890M Graphics (RADV GFX1150)" compute=0.0 id=00000000-c800-0000-0000-000000000000 pci_id=0000:c8:00.0
time=2025-11-05T18:51:54.786Z level=INFO source=server.go:400 msg="starting runner" cmd="/usr/bin/ollama runner --ollama-engine --port 46223"
time=2025-11-05T18:51:54.786Z level=DEBUG source=server.go:401 msg=subprocess PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin OLLAMA_FLASH_ATTENTION=true OLLAMA_DEBUG=1 HSA_OVERRIDE_GFX_VERSION=11.5.1 HSA_ENABLE_SDMA=0 OLLAMA_LLM_LIBRARY=vulkan LD_LIBRARY_PATH=/usr/lib/ollama:/usr/lib/ollama/vulkan:/usr/local/nvidia/lib:/usr/local/nvidia/lib64:/usr/lib/ollama:/usr/lib/ollama/rocm OLLAMA_HOST=0.0.0.0:11434 OLLAMA_LIBRARY_PATH=/usr/lib/ollama:/usr/lib/ollama/vulkan GGML_VK_VISIBLE_DEVICES=0 GGML_CUDA_INIT=1
time=2025-11-05T18:51:54.816Z level=DEBUG source=runner.go:471 msg="bootstrap discovery took" duration=29.909824ms OLLAMA_LIBRARY_PATH="[/usr/lib/ollama /usr/lib/ollama/vulkan]" extra_envs="map[GGML_CUDA_INIT:1 GGML_VK_VISIBLE_DEVICES:0]"
time=2025-11-05T18:51:54.816Z level=DEBUG source=runner.go:41 msg="GPU bootstrap discovery took" duration=200.314087ms
time=2025-11-05T18:51:54.816Z level=INFO source=types.go:42 msg="inference compute" id=00000000-c800-0000-0000-000000000000 filtered_id=0 library=Vulkan compute=0.0 name=Vulkan0 description="AMD Radeon 890M Graphics (RADV GFX1150)" libdirs=ollama,vulkan driver=0.0 pci_id=0000:c8:00.0 type=iGPU total="96.5 GiB" available="96.3 GiB"
```

```

total duration:       4m44.151979875s
load duration:        5.173844363s
prompt eval count:    170 token(s)
prompt eval duration: 790.302314ms
prompt eval rate:     215.11 tokens/s
eval count:           4902 token(s)
eval duration:        4m36.444275869s
eval rate:            17.73 tokens/s
root@ollama-docker-amd-230:~# docker exec ollama-vulkan-gpu ollama run --verbose gpt-oss:20b "Design a scalable web application architecture for an e-commerce platform that needs to handle 10,000 concurrent users during peak shopping periods. Your architecture should include a detailed database design with proper normalization, effective caching strategies using Redis or Memcached, load balancers, CDN integration, payment processing systems, inventory management, user authentication and authorization, session management, and monitoring tools. Explain the key components, their interactions, data flow between services, and how the system maintains performance and reliability under high traffic loads while ensuring secure transactions."
```

More later when the tests are done running.

jammsen commented:
Finally done 😅 

<img width="1557" height="528" alt="image" src="https://github.com/user-attachments/assets/e3600375-d413-4ace-8f99-561996224a7c" />

Vulkan doesnt look strong though.
Also did only AMD APU tests, took long enough 😎 and im not sure if the RTX2000E tests are really interesting for your comparison.

How does that 395+ run with theese models and ROCm vs Vulan @phueper - Can you cross compare a list with mine?

phueper commented:
> How does that 395+ run with theese models and ROCm vs Vulan @phueper - Can you cross compare a list with mine?

@jammsen  some results from my side, i am using https://github.com/phueper/ollama-benchmark to run them on some of the models that are available ...  as prompt i used the one you mentioned a few times here.

TLDR: 
- on the AI 5 340/840M  i currently have to force ROCm to work using the `HSA_OVERRIDE_GFX_VERSION` flag ... i didnt run rocm_v6 separately but results are basically the same as rocm_v7 only very slight differences. vulkan performs slightly better than rocm on this one
- on the AI MAX+ 395/8060S i would say the rocm and vulkan performance is very similar, some models quicker on one, some quicker on the other :shrug: 

## running on  AMD Ryzen AI 5 340 w/ Radeon 840M 

### rocm_v7 7.1.0 using `HSA_OVERRIDE_GFX_VERSION: "11.5.1"`

```shell
╰─$ dco --profile benchmark run --remove-orphans ollama-benchmark --models llama3 gemma3 gpt-oss --prompts "Design a scalable web application architecture for an e-commerce platform that needs to handle 10,000 concurrent users during peak shopping periods. Your architecture should include a detailed database design with proper normalization, effective caching strategies using Redis or Memcached, load balancers, CDN integration, payment processing systems, inventory management, user authentication and authorization, session management, and monitoring tools. Explain the key components, their interactions, data flow between services, and how the system maintains performance and reliability under high traffic loads while ensuring secure transactions." -t
```

```
Verbose: False
Test models: ['llama3', 'gemma3', 'gpt-oss']
Prompts: ['Design a scalable web application architecture for an e-commerce platform that needs to handle 10,000 concurrent users during peak shopping periods. Your architecture should include a detailed database design with proper normalization, effective caching strategies using Redis or Memcached, load balancers, CDN integration, payment processing systems, inventory management, user authentication and authorization, session management, and monitoring tools. Explain the key components, their interactions, data flow between services, and how the system maintains performance and reliability under high traffic loads while ensuring secure transactions.']
Table Output: True
Evaluating models: ['gemma3:4b', 'gpt-oss:20b', 'llama3:latest']

Table stats:
| Model         |            Prompt |   Evaluation |   Total |   Load Time |             Prompt |            Prompt |   Evalutaion |   Evaluation |   Total Time |
| Name          |   Evaluation Rate |         Rate |    Rate |         (s) |   Evaluation Count |   Evaluation Time |        Count |         Time |          (s) |
|               |             (T/s) |        (T/s) |   (T/s) |             |                    |               (s) |              |          (s) |              |
|---------------+-------------------+--------------+---------+-------------+--------------------+-------------------+--------------+--------------+--------------|
| gemma3:4b     |            104.98 |        10.33 |   11.12 |        5.32 |                116 |              1.10 |         1349 |       130.60 |       137.68 |
| gpt-oss:20b   |             95.09 |         6.54 |    6.75 |        8.97 |                170 |              1.79 |         4975 |       760.53 |       773.10 |
| llama3:latest |             89.76 |        11.09 |   12.12 |        3.81 |                113 |              1.26 |         1055 |        95.13 |       101.07 |
```

### vulkan

```shell
╰─$ dco --profile benchmark run --remove-orphans ollama-benchmark --models llama3 gemma3 gpt-oss --prompts "Design a scalable web application architecture for an e-commerce platform that needs to handle 10,000 concurrent users during peak shopping periods. Your architecture should include a detailed database design with proper normalization, effective caching strategies using Redis or Memcached, load balancers, CDN integration, payment processing systems, inventory management, user authentication and authorization, session management, and monitoring tools. Explain the key components, their interactions, data flow between services, and how the system maintains performance and reliability under high traffic loads while ensuring secure transactions." -t
```

```
Verbose: False
Test models: ['llama3', 'gemma3', 'gpt-oss']
Prompts: ['Design a scalable web application architecture for an e-commerce platform that needs to handle 10,000 concurrent users during peak shopping periods. Your architecture should include a detailed database design with proper normalization, effective caching strategies using Redis or Memcached, load balancers, CDN integration, payment processing systems, inventory management, user authentication and authorization, session management, and monitoring tools. Explain the key components, their interactions, data flow between services, and how the system maintains performance and reliability under high traffic loads while ensuring secure transactions.']
Table Output: True
Evaluating models: ['gemma3:4b', 'gpt-oss:20b', 'llama3:latest']

Table stats:
| Model         |            Prompt |   Evaluation |   Total |   Load Time |             Prompt |            Prompt |   Evalutaion |   Evaluation |   Total Time |
| Name          |   Evaluation Rate |         Rate |    Rate |         (s) |   Evaluation Count |   Evaluation Time |        Count |         Time |          (s) |
|               |             (T/s) |        (T/s) |   (T/s) |             |                    |               (s) |              |          (s) |              |
|---------------+-------------------+--------------+---------+-------------+--------------------+-------------------+--------------+--------------+--------------|
| gemma3:4b     |            130.12 |        17.87 |   19.09 |        2.75 |                116 |              0.89 |         1450 |        81.14 |        86.39 |
| gpt-oss:20b   |             62.35 |        13.59 |   13.97 |        7.87 |                170 |              2.73 |         4681 |       344.46 |       358.20 |
| llama3:latest |             82.60 |        12.94 |   14.32 |        3.70 |                113 |              1.37 |          873 |        67.47 |        73.26 |
```


## running on  AMD RYZEN AI MAX+ 395 w/ Radeon 8060S

### rocm_v7 7.1.0

```
Verbose: False
Test models: ['llama3', 'gemma3', 'gpt-oss']
Prompts: ['Design a scalable web application architecture for an e-commerce platform that needs to handle 10,000 concurrent users during peak shopping periods. Your architecture should include a detailed database design with proper normalization, effective caching strategies using Redis or Memcached, load balancers, CDN integration, payment processing systems, inventory management, user authentication and authorization, session management, and monitoring tools. Explain the key components, their interactions, data flow between services, and how the system maintains performance and reliability under high traffic loads while ensuring secure transactions.']
Table Output: True
Evaluating models: ['gemma3:12b', 'gemma3:270m', 'gemma3:4b', 'gpt-oss:120b', 'gpt-oss:20b', 'llama3.2:1b-instruct-q8_0', 'llama3.2:3b-instruct-q8_0', 'llama3.3:70b', 'llama3:latest']

Table stats:
| Model                     |            Prompt |   Evaluation |   Total |   Load Time |             Prompt |            Prompt |   Evalutaion |   Evaluation |   Total Time |
| Name                      |   Evaluation Rate |         Rate |    Rate |         (s) |   Evaluation Count |   Evaluation Time |        Count |         Time |          (s) |
|                           |             (T/s) |        (T/s) |   (T/s) |             |                    |               (s) |              |          (s) |              |
|---------------------------+-------------------+--------------+---------+-------------+--------------------+-------------------+--------------+--------------+--------------|
| gemma3:12b                |            358.95 |        22.42 |   23.79 |        2.78 |                116 |              0.32 |         1775 |        79.16 |        82.73 |
| gemma3:270m               |          12644.44 |       320.20 |  332.66 |        0.64 |                116 |              0.01 |         2903 |         9.07 |        10.57 |
| gemma3:4b                 |            813.01 |        63.81 |   68.93 |        1.89 |                116 |              0.14 |         1322 |        20.72 |        23.10 |
| gpt-oss:120b              |            350.48 |        30.90 |   31.60 |       15.10 |                170 |              0.49 |         6788 |       219.70 |       236.59 |
| gpt-oss:20b               |            744.78 |        44.35 |   45.73 |        5.44 |                170 |              0.23 |         5149 |       116.09 |       122.70 |
| llama3.2:1b-instruct-q8_0 |           2008.71 |       133.92 |  142.95 |        1.05 |                128 |              0.06 |         1763 |        13.16 |        14.90 |
| llama3.2:3b-instruct-q8_0 |           1009.99 |        54.80 |   59.94 |        1.79 |                128 |              0.13 |         1285 |        23.45 |        25.83 |
| llama3.3:70b              |             44.39 |         4.53 |    4.82 |       11.03 |                113 |              2.55 |         1556 |       343.86 |       358.05 |
| llama3:latest             |            405.36 |        44.63 |   49.29 |        2.11 |                113 |              0.28 |          951 |        21.31 |        24.05 |
```

### vulkan

```
Verbose: False
Test models: ['llama3', 'gemma3', 'gpt-oss']
Prompts: ['Design a scalable web application architecture for an e-commerce platform that needs to handle 10,000 concurrent users during peak shopping periods. Your architecture should include a detailed database design with proper normalization, effective caching strategies using Redis or Memcached, load balancers, CDN integration, payment processing systems, inventory management, user authentication and authorization, session management, and monitoring tools. Explain the key components, their interactions, data flow between services, and how the system maintains performance and reliability under high traffic loads while ensuring secure transactions.']
Table Output: True
Evaluating models: ['gemma3:12b', 'gemma3:270m', 'gemma3:4b', 'gpt-oss:120b', 'gpt-oss:20b', 'llama3.2:1b-instruct-q8_0', 'llama3.2:3b-instruct-q8_0', 'llama3.3:70b', 'llama3:latest']

Table stats:
| Model                     |            Prompt |   Evaluation |   Total |   Load Time |             Prompt |            Prompt |   Evalutaion |   Evaluation |   Total Time |
| Name                      |   Evaluation Rate |         Rate |    Rate |         (s) |   Evaluation Count |   Evaluation Time |        Count |         Time |          (s) |
|                           |             (T/s) |        (T/s) |   (T/s) |             |                    |               (s) |              |          (s) |              |
|---------------------------+-------------------+--------------+---------+-------------+--------------------+-------------------+--------------+--------------+--------------|
| gemma3:12b                |            314.37 |        24.08 |   25.68 |        3.85 |                116 |              0.37 |         1599 |        66.42 |        71.34 |
| gemma3:270m               |           3974.48 |       339.37 |  351.02 |        0.50 |                116 |              0.03 |         3081 |         9.08 |        10.52 |
| gemma3:4b                 |           1200.91 |        66.90 |   72.01 |        2.07 |                116 |              0.10 |         1427 |        21.33 |        24.05 |
| gpt-oss:120b              |            183.90 |        24.80 |   25.31 |       22.53 |                170 |              0.92 |         7090 |       285.92 |       311.59 |
| gpt-oss:20b               |            609.04 |        43.48 |   45.10 |        5.76 |                170 |              0.28 |         4223 |        97.12 |       104.36 |
| llama3.2:1b-instruct-q8_0 |           3898.09 |       138.65 |  150.26 |        1.06 |                128 |              0.03 |         1470 |        10.60 |        12.22 |
| llama3.2:3b-instruct-q8_0 |           1597.87 |        54.19 |   61.66 |        2.05 |                128 |              0.08 |          893 |        16.48 |        18.93 |
| llama3.3:70b              |             41.38 |         4.91 |    5.16 |       12.47 |                113 |              2.73 |         2001 |       407.33 |       423.41 |
| llama3:latest             |            661.47 |        42.21 |   47.42 |        1.33 |                113 |              0.17 |          849 |        20.11 |        21.92 |
```

jammsen commented:
Hey @phueper nice, slightly confused about the values - 2 Questions:
- Is evalrate or totalrate the Tokens per Second Rate that normally is referenced here?
- Also could you please share that script for benchmarking with me and also tell my how big the HDD-requirements are for alle the models to be downloaded? This can save me so much manual work, pleeeeeeeease share it 😭

phueper commented:
@jammsen  i laready did share it, look at the link in my message :) 
https://github.com/phueper/ollama-benchmark
it is a fork of an exisitng ollama-benchmark, ijust added the table layout of results and modifiied the filtering of models a bit 

i think the eval rate is what you usually see as a return from `ollama run --verbose`  the total rate is the the number of prompt eval tokens plus eval tokens divided by the total time 

and regarding the space usage:

```
$ du -hs files/ollama/storage/models 
179G    files/ollama/storage/models
```
or split up by the models i currenly have loaded (not all of them used for the benchmark run) 
```
╰─$ dco exec ollama ollama list                                                                                   
Alias tip: dce ollama ollama list
NAME                                  ID              SIZE      MODIFIED     
deepseek-coder:1.3b                   3ddd2d3fc8d2    776 MB    13 hours ago    
qwen2.5-coder:7b                      dae161e27b0e    4.7 GB    13 hours ago    
starcoder2:3b                         9f4ae0aff61e    1.7 GB    13 hours ago    
codegemma:2b                          926331004170    1.6 GB    13 hours ago    
qwen2.5-coder:3b                      f72c60cabf62    1.9 GB    13 hours ago    
gemma3:12b                            f4031aab637d    8.1 GB    13 hours ago    
deepseek-coder-v2:16b                 63fb193b3a9b    8.9 GB    13 hours ago    
gpt-oss:20b                           17052f91a42e    13 GB     13 hours ago    
mistral-nemo:latest                   e7e06d107c6c    7.1 GB    13 hours ago    
gemma3:270m                           e7d36fb2c3b3    291 MB    13 hours ago    
gemma3:4b                             a2af6cc3eb7f    3.3 GB    13 hours ago    
codellama:34b                         685be00e1532    19 GB     13 hours ago    
gpt-oss:120b                          a951a23b46a1    65 GB     13 hours ago    
llama3.2:3b-instruct-q8_0             e410b836fe61    3.4 GB    13 hours ago    
deepseek-r1:1.5b-qwen-distill-fp16    80c3a3c97773    3.6 GB    13 hours ago    
llama3.3:70b                          a6eb4748fd29    42 GB     13 hours ago    
llama3:latest                         365c0bd3c000    4.7 GB    13 hours ago    
llama3.2:1b-instruct-q8_0             baf6a787fdff    1.3 GB    13 hours ago
```

jammsen commented:
> @jammsen i laready did share it, look at the link in my message :) https://github.com/phueper/ollama-benchmark it is a fork of an exisitng ollama-benchmark, ijust added the table layout of results and modifiied the filtering of models a bit
> 
> i think the eval rate is what you usually see as a return from `ollama run --verbose` the total rate is the the number of prompt eval tokens plus eval tokens divided by the total time
> 
> and regarding the space usage:
> 
> ```
> $ du -hs files/ollama/storage/models 
> 179G    files/ollama/storage/models
> ```
> 
> or split up by the models i currenly have loaded (not all of them used for the benchmark run)
> 
> ```
> ╰─$ dco exec ollama ollama list                                                                                   
> Alias tip: dce ollama ollama list
> NAME                                  ID              SIZE      MODIFIED     
> deepseek-coder:1.3b                   3ddd2d3fc8d2    776 MB    13 hours ago    
> qwen2.5-coder:7b                      dae161e27b0e    4.7 GB    13 hours ago    
> starcoder2:3b                         9f4ae0aff61e    1.7 GB    13 hours ago    
> codegemma:2b                          926331004170    1.6 GB    13 hours ago    
> qwen2.5-coder:3b                      f72c60cabf62    1.9 GB    13 hours ago    
> gemma3:12b                            f4031aab637d    8.1 GB    13 hours ago    
> deepseek-coder-v2:16b                 63fb193b3a9b    8.9 GB    13 hours ago    
> gpt-oss:20b                           17052f91a42e    13 GB     13 hours ago    
> mistral-nemo:latest                   e7e06d107c6c    7.1 GB    13 hours ago    
> gemma3:270m                           e7d36fb2c3b3    291 MB    13 hours ago    
> gemma3:4b                             a2af6cc3eb7f    3.3 GB    13 hours ago    
> codellama:34b                         685be00e1532    19 GB     13 hours ago    
> gpt-oss:120b                          a951a23b46a1    65 GB     13 hours ago    
> llama3.2:3b-instruct-q8_0             e410b836fe61    3.4 GB    13 hours ago    
> deepseek-r1:1.5b-qwen-distill-fp16    80c3a3c97773    3.6 GB    13 hours ago    
> llama3.3:70b                          a6eb4748fd29    42 GB     13 hours ago    
> llama3:latest                         365c0bd3c000    4.7 GB    13 hours ago    
> llama3.2:1b-instruct-q8_0             baf6a787fdff    1.3 GB    13 hours ago
> ```

Awesome thanks will use that tomorrow! 
And im very sorry for missing the link, its 00:27am over here in Germany, sorry, need some sleep very tired. 😅

phueper commented:
fwiw ... i combined all changes from this one and #29  and created https://github.com/ollama/ollama/pull/13000   for upstream ... let's see if it will be considered :)
